src/main/kotlin/com/openlattice/BackgroundExternalDatabaseSyncingService.kt

```diff
@@ -0,0 +1,180 @@
+package com.openlattice
+
+import com.google.common.base.Stopwatch
+import com.hazelcast.core.HazelcastInstance
+import com.hazelcast.core.IMap
+import com.hazelcast.query.Predicate
+import com.hazelcast.query.Predicates
+import com.hazelcast.query.QueryConstants
+import com.openlattice.assembler.PostgresDatabases
+import com.openlattice.hazelcast.HazelcastMap
+import com.openlattice.indexing.MAX_DURATION_MILLIS
+import com.openlattice.indexing.configuration.IndexerConfiguration
+import com.openlattice.organization.OrganizationExternalDatabaseColumn
+import com.openlattice.organization.OrganizationExternalDatabaseTable
+import com.openlattice.organizations.ExternalDatabaseManagementService
+import org.apache.olingo.commons.api.edm.FullQualifiedName
+import org.slf4j.LoggerFactory
+import org.springframework.scheduling.annotation.Scheduled
+import java.util.*
+import java.util.concurrent.locks.ReentrantLock
+
+/**
+ *
+ */
+
+const val SCAN_RATE = 30_000L
+
+class BackgroundExternalDatabaseSyncingService(
+        private val hazelcastInstance: HazelcastInstance,
+        private val edms: ExternalDatabaseManagementService,
+        private val indexerConfiguration: IndexerConfiguration
+) {
+    companion object {
+        private val logger = LoggerFactory.getLogger(BackgroundExternalDatabaseSyncingService::class.java)
+    }
+
+    private val organizationExternalDatabaseColumns: IMap<UUID, OrganizationExternalDatabaseColumn> = hazelcastInstance.getMap(HazelcastMap.ORGANIZATION_EXTERNAL_DATABASE_COLUMN.name)
+    private val organizationExternalDatabaseTables: IMap<UUID, OrganizationExternalDatabaseTable> = hazelcastInstance.getMap(HazelcastMap.ORGANIZATION_EXTERNAL_DATABASE_TABLE.name)
+    private val aclKeys: IMap<String, UUID> = hazelcastInstance.getMap(HazelcastMap.ACL_KEYS.name)
+    private val organizationTitles: IMap<UUID, String> = hazelcastInstance.getMap(HazelcastMap.ORGANIZATIONS_TITLES.name)
+    private val expirationLocks: IMap<UUID, Long> = hazelcastInstance.getMap(HazelcastMap.EXPIRATION_LOCKS.name)
+
+
+    init {
+        expirationLocks.addIndex(QueryConstants.THIS_ATTRIBUTE_NAME.value(), true)
+    }
+
+    private val taskLock = ReentrantLock()
+
+    @Suppress("UNCHECKED_CAST", "UNUSED")
+    @Scheduled(fixedRate = MAX_DURATION_MILLIS)
+    fun scavengeExpirationLocks() {
+        expirationLocks.removeAll(
+                Predicates.lessThan(
+                        QueryConstants.THIS_ATTRIBUTE_NAME.value(),
+                        System.currentTimeMillis()
+                ) as Predicate<UUID, Long>
+        )
+    }
+
+    @Suppress("UNUSED")
+    @Scheduled(fixedRate = SCAN_RATE)
+    fun scanOrganizationDatabases() {
+        logger.info("Starting background external database sync task.")
+        if (taskLock.tryLock()) {
+            try {
+                if (indexerConfiguration.backgroundExternalDatabaseSyncingEnabled) {
```

Comment:
I would move this up to the top of this function to avoid the need to be constantly locking/unlocking. I would also flip around the if/else in order to reduce nesting. i.e.
```
if (!indexerConfiguration.backgroundExternalDatabaseSyncingEnabled) {
  logger.info("Skipping external database syncing as it is not enabled.")
  return
}

if (!taskLock.tryLock()) {
  logger.info("Not starting new external database sync task as an existing one is running")
  return
}

try {
  ...blah...
} finally {
  taskLock.unlock()
}
```